newsroom/history_async.py

```diff
@@ -0,0 +1,184 @@
+            pass
+
+    async def query_items(self, query: Dict[str, Any]):
+        if query["from"] >= 1000:
+            # https://www.elastic.co/guide/en/elasticsearch/guide/current/pagination.html#pagination
+            abort(400)
+
+        # Use self.find to execute the query and get the cursor
+        cursor = await self.find(query)
+        return cursor
+
+    async def fetch_history(self, query: Dict[str, Any], all: bool = False):
+        cursor = await self.query_items(query)
+
+        # Fetch the documents from the cursor
+        docs = []
+        async for doc in cursor:
+            docs.append(doc)
+
+        if all:
+            # Handle pagination and retrieve additional results
+            while cursor.hits["hits"]["total"]["value"] > len(docs):
+                query["from"] = len(docs)
+                cursor = await self.query_items(query)
+                async for doc in cursor:
+                    docs.append(doc)
+
+        # Return the results
+        return {"_items": docs, "hits": cursor.hits}
+
+
+async def get_history_users(
+    item_ids: list[ObjectId | str],
+    active_user_ids: list[ObjectId | str],
+    active_company_ids: list[ObjectId | str],
+    section: str,
+    action: str,
+) -> List[str]:
+    source = {
+        "query": {
+            "bool": {
+                "filter": [
+                    {
+                        "bool": {
+                            "should": [
+                                {"terms": {"company": [str(a) for a in active_company_ids]}},
+                                {"bool": {"must_not": [{"exists": {"field": "company"}}]}},
+                            ],
+                            "minimum_should_match": 1,
+                        },
+                    },
+                    {"terms": {"item": [str(i) for i in item_ids]}},
+                    {"term": {"section": section}},
+                    {"term": {"action": action}},
+                ]
+            }
+        },
+        "size": 25,
+        "from": 0,
+    }
+
+    histories_cursor = await HistoryService().find(source)
+
+    # Collect the history items
+    histories_items = []
+    async for history in histories_cursor:
+        histories_items.append(history)
+
+    # Filter out the users
+    user_ids = [str(uid) for uid in active_user_ids]
+    return [str(h["user"]) for h in histories_items if str(h["user"]) in user_ids]
+
+
+history_resource_config = ResourceConfig(
+    name="history",
+    data_class=HistoryResourceModel,
+    service=HistoryService,
+    mongo=MongoResourceConfig(
+        prefix=MONGO_PREFIX,
+        indexes=[
+            MongoIndexOptions(
+                name="item",
+                keys=[("item", 1)],
+            ),
+            MongoIndexOptions(
+                name="company_user",
+                keys=[("company", 1), ("user", 1)],
+            ),
+        ],
+    ),
+    elastic=ElasticResourceConfig(),
```

Comment:
You'll have to set a prefix here from `newsroom.ELASTIC_PREFIX` value. That will fix the index issues